ddtrace/internal/_threads.cpp

```diff
@@ -431,9 +428,33 @@ PeriodicThread__atexit(PeriodicThread* self, PyObject* args)
 
 // ----------------------------------------------------------------------------
 static PyObject*
-PeriodicThread__after_fork(PeriodicThread* self, PyObject* args)
+PeriodicThread__after_fork(PeriodicThread* self)
+{
+    self->_thread = nullptr;
+
+    self->_stopping = false;
+    self->_atexit = false;
+    self->_skip_shutdown = false;
+
+    self->_started->clear();
+    self->_stopped->clear();
+    self->_request->clear();
+    self->_served->clear();
+
+    PeriodicThread_start(self);
+
+    Py_RETURN_NONE;
+}
+
+// ----------------------------------------------------------------------------
+static PyObject*
+PeriodicThread__before_fork(PeriodicThread* self)
 {
-    self->_after_fork = true;
+    self->_skip_shutdown = true;
+
+    PeriodicThread_stop(self);
+
+    self->_thread = nullptr;
```

Comment:
In `PeriodicThread_start` we detach the thread, so as far as I can tell destructing the `std::thread` here won't synchronize. Nor will `PeriodicThread_stop`. Do we need synchronization here to make sure the thread is actually stopped before forking? Like, on `self->_stopped` as we do in `PeriodicThread_join`? (or just calling that method?)

Reply:
We call `join` in `threads.py::_before_fork` to ensure that the thread has actually terminated

Reply:
Correct!